cpp/src/arrow/filesystem/gcsfs.cc

```diff
@@ -145,6 +189,25 @@ class GcsFileSystem::Impl {
     return std::make_shared<GcsInputStream>(std::move(stream));
   }
 
+  Result<std::shared_ptr<io::OutputStream>> OpenOutputStream(
+      const GcsPath& path, const std::shared_ptr<const KeyValueMetadata>& metadata) {
+    gcs::EncryptionKey encryption_key;
+    ARROW_ASSIGN_OR_RAISE(encryption_key, internal::ToEncryptionKey(metadata));
+    gcs::PredefinedAcl predefined_acl;
+    ARROW_ASSIGN_OR_RAISE(predefined_acl, internal::ToPredefinedAcl(metadata));
+    gcs::KmsKeyName kms_key_name;
+    ARROW_ASSIGN_OR_RAISE(kms_key_name, internal::ToKmsKeyName(metadata));
+    gcs::WithObjectMetadata with_object_metadata;
+    ARROW_ASSIGN_OR_RAISE(with_object_metadata, internal::ToObjectMetadata(metadata));
+
+    auto stream = client_.WriteObject(path.bucket, path.object, encryption_key,
+                                      predefined_acl, kms_key_name, with_object_metadata);
+    if (!stream.last_status().ok()) {
+      return internal::ToArrowStatus(stream.last_status());
+    }
```

Comment:
You might want to define a macro that would allow you to write:
```c++
ARROW_GCS_RETURN_NOT_OK(stream.last_status());
```

Reply:
Done.